<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Guided-Capstone-Step-4---Pre-processing-and-Training-Data-Development" data-toc-modified-id="Guided-Capstone-Step-4---Pre-processing-and-Training-Data-Development-0"><span class="toc-item-num">0&nbsp;&nbsp;</span>Guided Capstone Step 4 - Pre-processing and Training Data Development</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#The-Data-Science-Method" data-toc-modified-id="The-Data-Science-Method-0.0.1"><span class="toc-item-num">0.0.1&nbsp;&nbsp;</span><strong>The Data Science Method</strong></a></span></li></ul></li><li><span><a href="#Create-dummy-features-for-categorical-variables" data-toc-modified-id="Create-dummy-features-for-categorical-variables-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Create dummy features for categorical variables</a></span></li><li><span><a href="#Standardize-the-magnitude-of-numeric-features" data-toc-modified-id="Standardize-the-magnitude-of-numeric-features-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>Standardize the magnitude of numeric features</a></span></li><li><span><a href="#Split-into-training-and-testing-datasets" data-toc-modified-id="Split-into-training-and-testing-datasets-0.3"><span class="toc-item-num">0.3&nbsp;&nbsp;</span>Split into training and testing datasets</a></span></li></ul></li><li><span><a href="#Guided-Capstone-Step-5---Modeling" data-toc-modified-id="Guided-Capstone-Step-5---Modeling-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Guided Capstone Step 5 - Modeling</a></span><ul class="toc-item"><li><span><a href="#Fit-Models-with-Training-Data-Set" data-toc-modified-id="Fit-Models-with-Training-Data-Set-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Fit Models with Training Data Set</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Model-1" data-toc-modified-id="Model-1-1.1.0.1"><span class="toc-item-num">1.1.0.1&nbsp;&nbsp;</span>Model 1</a></span></li></ul></li></ul></li><li><span><a href="#Review-Model-Outcomes-—-Iterate-over-additional-models-as-needed." data-toc-modified-id="Review-Model-Outcomes-—-Iterate-over-additional-models-as-needed.-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Review Model Outcomes — Iterate over additional models as needed.</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Model-2" data-toc-modified-id="Model-2-1.2.0.1"><span class="toc-item-num">1.2.0.1&nbsp;&nbsp;</span>Model 2</a></span></li><li><span><a href="#Model-3" data-toc-modified-id="Model-3-1.2.0.2"><span class="toc-item-num">1.2.0.2&nbsp;&nbsp;</span>Model 3</a></span></li></ul></li></ul></li><li><span><a href="#Identify-the-Final-Model" data-toc-modified-id="Identify-the-Final-Model-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Identify the Final Model</a></span></li></ul></li></ul></div>

# Guided Capstone Step 4 - Pre-processing and Training Data Development

### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   **Pre-processing and Training Data Development**

5.  Modeling
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

In this step of the guided capstone, you'll revisit some of the steps you completed to solve questions in step 4. Revisiting these steps will help you get additional practice completing this kind of work.

**<font color='teal'> Start by loading the necessary packages, as you've done in the previous steps. Print out your current working directory to confirm that you are in the correct project directory. </font>**

In [146]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
from sklearn import metrics
from sklearn.preprocessing import scale
from sklearn.model_selection import train_test_split

**<font color='teal'>  Load the csv file you created in step 3. Remember, it should be saved inside your data subfolder. Print the first five rows.</font>**

In [147]:
df = pd.read_csv('data/step3_output.csv')
print(df.head())
df.info()

                            Name       state  vertical_drop  base_elev  trams  \
0                 Alyeska Resort      Alaska           2500        250      1   
1               Hilltop Ski Area      Alaska            294       1796      0   
2               Arizona Snowbowl     Arizona           2300       9200      0   
3            Sunrise Park Resort     Arizona           1800       9200      0   
4  Yosemite Ski & Snowboard Area  California            600       7200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  \
0        0.0          0          2     2       0  ...             1610.0   
1        0.0          0          0     0       1  ...               30.0   
2        0.0          1          0     2       2  ...              777.0   
3        0.0          0          1     2       3  ...              800.0   
4        0.0          0          0     0       1  ...               88.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageS

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `State`. Add the dummies back to the dataframe and remove the original column for `State`. </font>**

In [148]:
df = pd.concat([df.drop('state', axis=1), pd.get_dummies(dfo)], axis=1)
print(df.head())

                            Name  vertical_drop  base_elev  trams  fastEight  \
0                 Alyeska Resort           2500        250      1        0.0   
1               Hilltop Ski Area            294       1796      0        0.0   
2               Arizona Snowbowl           2300       9200      0        0.0   
3            Sunrise Park Resort           1800       9200      0        0.0   
4  Yosemite Ski & Snowboard Area            600       7200      0        0.0   

   fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0          0          2     2       0       0  ...             0   
1          0          0     0       1       0  ...             0   
2          1          0     2       2       1  ...             0   
3          0          1     2       3       1  ...             0   
4          0          0     0       1       3  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0  

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing, standardize the scale of the features of the dataframe except the name of the resort, which you don't need in the dataframe for modeling so it can be droppped here as well. You should  hold out your response variable(s) so that you can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as your response for scaling and modeling. Later, you will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [149]:
# Remove column Name from the dataframe
df.drop('Name', axis=1, inplace=True)

In [150]:
# Assign Y to AdultWeekend and the rest of the dataframe to X
Y = df['AdultWeekend']
X = df.drop('AdultWeekend', axis=1)

In [151]:
# Standardize the scale of each features in X
data = scale(X, axis=0, with_mean=True, with_std=True, copy=True)


## Split into training and testing datasets

**<font color='teal'> Using sklearn model selectionm import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. You will start by using the adult weekend ticket price as your response variable for modeling.</font>**

In [152]:
# Create a 75/25 train/test split
X_train, X_test, y_train, y_test = train_test_split(data, Y, test_size=0.25, random_state=42)

Here, you'll start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5 - Modeling

You've cleaned and prepared the datasets, so now it's time to get into the most exciting part of this process: modeling! In this exercise, you will build three different models and compare each model's performance. In the end, you will choose the best model for proving insights to Big Mountain management.  

1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with Training Data Set

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [153]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [154]:
# Create a prediction y_pred
y_pred = lm.predict(X_test)

In [155]:
print('Explained Variance Score (R-squared): %.2f'
      % explained_variance_score(y_test, y_pred))
# The coefficient of determination: 1 is perfect prediction
print('Mean Absolute Error: %.2f'
      % mean_absolute_error(y_test, y_pred))

Explained Variance Score (R-squared): 0.80
Mean Absolute Error: 5.09


## Review Model Outcomes — Iterate over additional models as needed.

In [156]:
#Intercept value from the linear model
lm.intercept_

61.07156049389448

**<font color='teal'> Print the intercept value from the linear model. </font>**

In [157]:
# Sort and print coeff
coeff_df = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
print(coeff_df.sort_values(by=['Coefficient'], ascending=False))

# The chairs columns made it to the top of my list. Is it because if Step 3 of my choice of replacing null value by zero or mean()?

                    Coefficient
total_chairs       2.756926e+12
surface            1.002605e+12
double             9.703812e+11
fastQuads          9.563704e+11
triple             8.700398e+11
quad               6.874604e+11
New York           6.309010e+11
Michigan           5.836520e+11
Pennsylvania       4.803653e+11
Wisconsin          4.529635e+11
New Hampshire      4.529635e+11
Colorado           4.384418e+11
California         4.384418e+11
Vermont            4.232986e+11
Minnesota          4.074646e+11
Montana            3.908558e+11
Massachusetts      3.908558e+11
Idaho              3.548740e+11
Washington         3.548740e+11
Utah               3.352044e+11
New Mexico         3.141394e+11
Wyoming            3.141394e+11
Oregon             3.141394e+11
fastSixes          3.105575e+11
Maine              2.913767e+11
Ohio               2.664814e+11
North Carolina     2.664814e+11
Virginia           2.387876e+11
West Virginia      2.387876e+11
trams              2.375225e+11
Iowa    

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the `AdultWeekend` resulting value. Also, because you took the time to scale your x values in the training data, you can compare each of the coefficients for the features to determine the feature's importance. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what you are looking for is the magnitude of impact on your response variable.

In [158]:
# Pass X_train & X_test as DataFrame and add columns names
X_train_df = pd.DataFrame(X_train, columns=X.columns)
X_test_df = pd.DataFrame(X_test, columns=X.columns)


# Drop the states columns
X_train = X_train_df.drop(X_train_df.iloc[:,23:], axis = 1, inplace = False)
X_test = X_test_df.drop(X_test_df.iloc[:,23:], axis = 1, inplace = False)

The result as not as good as model one as the location of the resort plays a big importance on the price.

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that you care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

#### Model 2

In [159]:
#model2
model2 = lm2.fit(X_train,y_train)

# Create a prediction y_pred
y_pred2 = lm2.predict(X_test)

print('Explained Variance Score (R-squared): %.2f'
      % explained_variance_score(y_test, y_pred2))
# The coefficient of determination: 1 is perfect prediction
print('Mean Absolute Error: %.2f'
      % mean_absolute_error(y_test, y_pred2))

Explained Variance Score (R-squared): 0.75
Mean Absolute Error: 6.04


In [160]:
# Sort and print coeff
coeff_df = pd.DataFrame(abs(lm2.coef_), X_test.columns, columns=['Coefficient'])
print(coeff_df.sort_values(by=['Coefficient'], ascending=False))

                   Coefficient
AdultWeekday         12.607115
vertical_drop         2.705003
SkiableTerrain_ac     1.781336
projectedDaysOpen     1.503339
trams                 1.440023
daysOpenLastYear      1.349606
quad                  1.274282
NightSkiing_ac        1.243065
kmean_cluster         1.183630
base_elev             1.087776
Runs                  0.977167
triple                0.610132
fastEight             0.556737
LongestRun_mi         0.541001
fastQuads             0.470281
total_chairs          0.434284
surface               0.431172
averageSnowfall       0.329976
Snow Making_ac        0.323251
fastSixes             0.226599
yearsOpen             0.200986
double                0.064247
TerrainParks          0.019837


**<font color='teal'> When reviewing your new model coefficients, note that `summit_elev` is now in the number two spot. From a managerial perspective, this is also difficult to change and highly correlated with `base_elev` and `vertical_drop`.  This time rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [161]:
# summit_elev was removed in step3 when checking high correlation between 2 features.
# Remove base_elev
X_train = X_train.drop(['base_elev'], axis = 1)
X_test = X_test.drop(['base_elev'], axis = 1)


#model3 
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X_train,y_train)

# Create a prediction y_pred
y_pred3 = lm3.predict(X_test)

print('Explained Variance Score (R-squared): %.2f'
      % explained_variance_score(y_test, y_pred3))
# The coefficient of determination: 1 is perfect prediction
print('Mean Absolute Error: %.2f'
      % mean_absolute_error(y_test, y_pred3))

Explained Variance Score (R-squared): 0.75
Mean Absolute Error: 6.08


In [162]:
# Sort and print coeff
coeff_df = pd.DataFrame(abs(lm3.coef_), X_test.columns, columns=['Coefficient'])
print(coeff_df.sort_values(by=['Coefficient'], ascending=False))

                   Coefficient
AdultWeekday         12.432872
vertical_drop         2.904696
kmean_cluster         1.833248
SkiableTerrain_ac     1.688649
NightSkiing_ac        1.467509
quad                  1.322176
trams                 1.312774
projectedDaysOpen     1.304340
daysOpenLastYear      1.136212
Runs                  0.661099
triple                0.562092
LongestRun_mi         0.558136
averageSnowfall       0.543042
fastEight             0.525908
total_chairs          0.465281
fastQuads             0.402880
surface               0.375624
fastSixes             0.320867
Snow Making_ac        0.218045
yearsOpen             0.207519
double                0.108164
TerrainParks          0.036461


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell. You will explain your selection during the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.80 |5.09  |-|
| Model 2. | 0.75|6.04 |'state'|
| Model 3. | 0.75 |6.08 |'state','summit_elev','base_elev'|

Model Selection:

Regardless of the performance of the model I will choose model 3 because we want to define a new lift ticket price, regardless of the location and altitude of the resort.